indexer/indexer.py

```diff
@@ -0,0 +1,337 @@
+        The constructor creates an Indexer instance, with various
+        parameters that are needed when performing the indexing operation.
+        It requires a dictionary of the metadata_files, the data_files,
+        an ElasticSearch(ES) client object to communicate with some
+        ES instance, the index_settings, as well as a mapping config
+        file that can be used by the indexer for indexing operation and
+        actual ElasticSearch mapping. Any extra kwargs are set as
+        attributes to the instance.
+
+        :param metadata_files: dictionary of json metadata (list of dicts)
+        :param data_files: dictionary describing non-metadata files
+         (list of dicts)
+        :param es_client: The elasticsearch client
+        :param index_name: The name of the index to put the documents.
+        :param doc_type: The document type to put the document under in ES.
+        :param index_settings: Any special settings for the ES index.
+        :param index_mapping_config: The indexer config file.
+        """
+        # Set main arguments
+        self.metadata_files = metadata_files
+        self.data_files = data_files
+        self.es_client = es_client
+        self.index_name = index_name
+        self.doc_type = doc_type
+        self.index_settings = index_settings
+        self.index_mapping_config = index_mapping_config
+        # Set all kwargs as instance attributes
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def index(self, bundle_uuid, bundle_version, *args, **kwargs):
+        """
+        Indexes the data files.
+
+        Triggers the actual indexing process
+
+        :param bundle_uuid: The bundle_uuid of the bundle that will be indexed.
+        :param bundle_version: The bundle of the version.
+        """
+        raise NotImplementedError(
+            'users must define index to use this base class')
+
+    def special_fields(self, *args, **kwargs):
+        """
+        Add any special fields that may be missing.
+
+        Gets any special field that may not be available directly from the
+        metadata.
+
+        :return: a dictionary of all the special fields to be added.
+        """
+        raise NotImplementedError(
+            'users must define special_fields to use this base class')
+
+    def merge(self, doc_contents, **kwargs):
+        """
+        Merge the document with the contents in ElasticSearch.
+
+        merge() should take the results of get_item() and harmonize it
+        with whatever is present in ElasticSearch to avoid blind overwritting
+        of documents. Users should implement their own protocol.
+
+        :param doc_contents: Current document to be indexed.
+        :return: The harmonized document which can overwrite existing entry.
+        """
+        raise NotImplementedError(
+            'users must define merge to use this base class')
+
+    def load_doc(self, doc_uuid, doc_contents, * args, **kwargs):
```

Comment:
This seems to be a helper method for subclasses. It should be protected (name starts with `_`).

Reply:
Name should be `save_doc`. See my other comment on this naming issue.